### PR TITLE
[libc++] Refactor some code in monotonic_buffer_resource

### PR DESCRIPTION
**Pull request by love1angel**
1. remove unused __default_buffer_alignment
2. two __try_allocate_from_chunk are same, put it together

This patch refactor some code in monotonic_buffer_resource.

### CODE REVIEW

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (frederick-vs-ja) [inline libcxx/include/__memory_resource/monotonic_buffer_resource.h:31]: `__default_growth_factor` is not used in headers. I _guess_ it would be better to move it to `memory_resource.cpp`, but I'd ask maintainers first.

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (frederick-vs-ja) [inline libcxx/include/__memory_resource/monotonic_buffer_resource.h:41]: Oops, both `__try_allocate_from_chunk` functions are ABI-critical, which is probably indicated by lack of `_LIBCPP_HIDE_FROM_ABI`. They must still exist, although I'm not sure whether it's possibly to change their behaivor (if so, they can just return `nullptr`).

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (ldionne) [inline libcxx/include/__memory_resource/monotonic_buffer_resource.h:31]: Was `__default_buffer_alignment` unused?

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (ldionne) [reply libcxx/include/__memory_resource/monotonic_buffer_resource.h:41]: Those functions are on the ABI boundary, we can't remove them or change their behavior in a backwards incompatible way because pre-existing programs compiled against older versions of the headers will contain calls to this external function.

Unless nothing in the headers was ever calling into this function, in which case we might be able to remove it from the dylib. Is that the situation?

Edit: Doing a bit of searching, it seems like nobody's depending on that symbol, which would support that it was actually never needed as part of the ABI. I think it would be OK to remove it. You will need to update the `libcxx/lib/abi/*.abilist` files and `libcxx/lib/abi/CHANGELOG.TXT`.

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (ldionne) [inline libcxx/include/__memory_resource/monotonic_buffer_resource.h:53]: Can this be moved to a non-member function defined only in `memory_resource.cpp`?

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (love1angel) [reply libcxx/include/__memory_resource/monotonic_buffer_resource.h:41]: thanks, i will change

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (love1angel) [reply libcxx/include/__memory_resource/monotonic_buffer_resource.h:53]: > Can this be moved to a non-member function defined only in `memory_resource.cpp`?

since struct __chunk_footer is private inside class monotonic_buffer_resource. then i couldn't get scope struct __chunk_footer __initial_descriptor.

but this can be easy archieved by add a branch judge cost.  otherwise i can distinguish them by struct size, but this is ugly.

Comment (frederick-vs-ja) [reply libcxx/include/__memory_resource/monotonic_buffer_resource.h:53]: On the "ugliness": Oh, currently non-__ugllified identifiers are introduced in a header. If you want to add some implementation details to a header, the identifiers should be __uglified.

> but this can be easy archieved by add a branch judge cost.

I think the cost would be paid at only build-time - it should be just adding and manually specifying another template argument.

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (frederick-vs-ja) [reply libcxx/include/__memory_resource/monotonic_buffer_resource.h:31]: It seems that `__default_buffer_alignment` has never been used since intiallly added (243da90ea5357c1ca324f714ea4813dc9029af27, https://reviews.llvm.org/D89057).

File: libcxx/include/__memory_resource/monotonic_buffer_resource.h
Comment (love1angel) [reply libcxx/include/__memory_resource/monotonic_buffer_resource.h]: done, thanks

File: libcxx/src/memory_resource.cpp
Comment (frederick-vs-ja) [inline libcxx/src/memory_resource.cpp:432]: I think the explicitly specified template parameter should appear first.

```suggestion
template <bool is_initial, typename Chunk>
```

The usages below should be modified correspondingly.

Comment (love1angel) [reply libcxx/src/memory_resource.cpp]: > I think the explicitly specified template parameter should appear first.
> 
> The usages below should be modified correspondingly.

done

**Review by ldionne — COMMENTED** [review]
Thanks a lot for the patch! I think this makes sense and we'll be able to move this forward, but I left some questions and comments.

File: libcxx/src/memory_resource.cpp
Comment (love1angel) [reply libcxx/src/memory_resource.cpp]: done

**Review by ldionne — APPROVED** [review]
LGTM pending green CI!

Thanks for the refactoring!

**Review by frederick-vs-ja — APPROVED** [review]
CI should be fixed now. I'm updating the branch to make CI green.

**Pull request merged**
